Adding tags to resources #465 (#232)
alphagov/govuk-infrastructure: terraform/docs/tagging-guide.md

# Overview
This document will describe the tagging strategy to be used within Replatforming terraform code which will be used to create resources in the AWS environment infrastructure.

Tagging is required so that cost and resource utilisation can be processed, with an added benefit of AWS console **Tag Key** searching.

# Tagging Table

The following table highlights the tags **Tag Key** that should added to the AWS resources listed in the section "AWS Resources that can be tagged".
The common column distinguishes between Tags that have been added as part of an default set and represented by **yes** and those which are unique per service represented by **no**

| **Tag Key** | **Tag Value(s)** | **Description** | **Example** | **Common** |
|--|--|--|--|--|
| Name | [ServiceName]-[Environment]-[Workspace] | This is the identifiable name of the service. | publisher-test-nadeem | no |
| chargeable_entity | govuk-publishing-platform-[Environment] | This is required for billing. | govuk-publishing-platform-test | yes |

> | environment | test int stag prod | Environment area to which this belongs. | test | yes |

Reviewer: (just making these consistent with the names as typically used elsewhere in GOV.UK)
```suggestion
| environment | test integration staging production | Environment area to which this belongs. | test | yes |

```

Author: thank you. totally forgot to update the doc

|project | replatforming | This is the project under which this was developed. | replatforming | yes |
| repository | govuk-aws govuk-infrastructure | This is the Git repo in which this service resides. | govuk-infrastructure | yes |
| terraform_deployment | concourse-iam govuk-publishing-platform task-runner monitoring-test | The source in which the service resides. | govuk-publishing-platform | yes | 

> |terraform_workspace | bill chris fred karl nadeem steve roch towers | This should be the name of the terraform workspace that created the service. | nadeem | yes |

Reviewer: (let's add `default` as an example since it's the most important one)
```suggestion
|terraform_workspace | default bill chris fred karl nadeem steve roch towers | This should be the name of the terraform workspace that created the service. | nadeem | yes |

```

```
tags = merge(
    local.additional_tags,
    {
      Name = "publisher-${var.environment}-${local.workspace}"
    }
```
- Below example of a correct module tag definition for the non common **Tag Key** Name 
```
tags = merge(
    var.additional_tags,
    {
      Name = "publisher-${var.environment}-${var.workspace}"
    }
```

- Below example of local common tags defined in the main.tf file
```
locals {
  additional_tags = {
    chargeable_entity    = "govuk-publishing-platform-${var.govuk_environment}"
    environment          = var.govuk_environment
    project              = "replatforming"
    repository           = "govuk-infrastructure"
    terraform_deployment = "govuk-publishing-platform"
    terraform_workspace  = terraform.workspace
  }
}
```

**IMPORTANT :-** 
- The **Key** attribute **Name** should start with an Uppercase letter and the rest should be lowercase with no spaces. 
- The **Value** attribute should be lowercase and no spaces however hyphens can be used.

**NOTES :-** 
- All listed resources from below should be made compliant.
- Common Tags have been added as **locals** with in the deployment terraform **main** file.

> - This tagging strategy should ideally be replicated to other and new yet to be deployed environments such as **int**.

Reviewer: ```suggestion
- This tagging strategy should ideally be replicated to other and new yet to be deployed environments such as **integration**.

```

Author: Yes i forget to update the Doc with environment names - will do a git amend
